service/s3/marshal.go

```diff
@@ -0,0 +1,162 @@
+		encoder.AddHeader("x-amz-server-side-encryption-context").String(*input.SSEKMSEncryptionContext)
+	}
+	if input.SSEKMSKeyId != nil {
+		encoder.AddHeader("x-amz-server-side-encryption-aws-kms-key-id").String(*input.SSEKMSKeyId)
+	}
+	if input.SSECustomerKey != nil {
+		encoder.AddHeader("x-amz-server-side-encryption-customer-key-MD5").String(*input.SSECustomerKeyMD5)
+	}
+	if input.SSECustomerKeyMD5 != nil {
+		encoder.AddHeader("x-amz-server-side-encryption-customer-key-MD5").String(*input.SSECustomerKeyMD5)
+	}
+	if input.SSECustomerAlgorithm != nil {
+		encoder.AddHeader("x-amz-server-side-encryption-customer-algorithm").String(*input.SSECustomerAlgorithm)
+	}
+	if input.WebsiteRedirectLocation != nil {
+		encoder.AddHeader("x-amz-website-redirect-location").String(*input.WebsiteRedirectLocation)
+	}
+	if input.StorageClass != "" {
+		encoder.AddHeader("x-amz-storage-class").String(string(input.StorageClass))
+	}
+	if input.ServerSideEncryption != "" {
+		encoder.AddHeader("x-amz-server-side-encryption").String(string(input.ServerSideEncryption))
+	}
+	if input.Expires != nil {
+		if err := encoder.AddHeader("Expires").Time(*input.Expires, protocol.RFC822TimeFormatName); err != nil {
+			return awserr.New(aws.ErrCodeSerialization, "failed to marshal header for shape Expires using REST encoder:\n \t %v", err)
+		}
+	}
+	if input.ObjectLockRetainUntilDate != nil {
+		if err := encoder.AddHeader("x-amz-object-lock-retain-until-date").Time(*input.ObjectLockRetainUntilDate, protocol.ISO8601TimeFormatName); err != nil {
+			return awserr.New(aws.ErrCodeSerialization, "failed to marshal header for shape ObjectLockRetainUntilDate using REST encoder:\n \t %v", err)
+		}
+	}
+	//  Encoding shapes with location `uri`
+	if input.Bucket != nil {
+		if err := encoder.SetURI("Bucket").String(*input.Bucket); err != nil {
+			return awserr.New(aws.ErrCodeSerialization, "failed to marshal URI for shape Bucket using REST encoder:\n \t %v", err)
+		}
+	}
+	if input.Key != nil {
+		if err := encoder.SetURI("Key").String(*input.Key); err != nil {
+			return awserr.New(aws.ErrCodeSerialization, "failed to marshal URI for shape Key using REST encoder:\n \t %v", err)
+		}
+	}
+	return nil
+}
+
+// marshalPutObjectInputShapeAWSXML is a stand alone function used to marshal the xml payload
+// This should be generated according to the payload type for rest-xml protocol
+func marshalPutObjectInputShapeAWSXML(r *aws.Request, input *PutObjectInput) error {
+	if input.Body != nil {
+		r.SetReaderBody(input.Body)
+	}
+	return nil
+}
+
+// marshalShapeMapForHeaders is marshal function that takes in a map[string]string as an input along with an encoder
+// and location Name which should be used to marshal the shape with location headers.
+func marshalShapeMapForHeaders(encoder *rest.Encoder, locationName string, input map[string]string) {
+	headerObject := encoder.Headers(locationName)
+	for k, v := range input {
+		headerObject.AddHeader(k).String(v)
+	}
+}
```

Comment:
Not as important for headers but, when the SDK starts generating XML and JSON objects from `map` and `slices`. Specifically the handling around empty vs nil maps/slices.  A empty, but not nil map/slice would be serialized as an empty JSON object/list. Whereas if the map/slice were nil nothing would be marshaled.